zschema/keys.py

```diff
@@ -30,6 +30,31 @@ def __cmp__(self, other):
             return cmp(int(self.port), int(other.port))
 
 
+# Create a dict with all the keys / values of rhs and lhs, where values
+# from lhs are used if a key is in both.
+def left_merge(lhs, rhs):
+    ret = rhs.copy()
+    ret.update(lhs)
+    return ret
+
+# Factory for TypeFactorys (e.g. SubRecordType is a TypeFactory for
+# SubRecords).
+class TypeFactoryFactory(object):
+    # This factory acts as a constructor for cls with the first 
+    # len(args) positional arguments fixed by args, and with kwargs 
+    # serving as the default keyword arguments.
+    def __init__(self, cls, args, kwargs):
+        self.args = args
+        self.kwargs = kwargs
+        self.cls = cls
+
+    # Returns a TypeFactory that returns type instances of cls, using
+    # the args positional arguments, and using kwargs as the default
+    # values for any keyword arguments.
+    def __call__(self, *args, **kwargs):
+        return self.cls(*(self.args + args), **left_merge(kwargs, self.kwargs))
```

Comment:
Aaahhh that list addition is scary. That seems ripe for error. @justinbastress I do agree with your prior comment. What if we allowed args in either place but not both?

Reply:
I'd agree that the appending behavior is less obvious (and so prone to mistakes).

Allowing it in either/or should cover most the feasible uses and virtually all of the practical uses (hence the need to contrive that `class Positional` in the tests), so that sounds like a reasonable way to prevent mistakes without losing any real functionality.